node/core/approval-voting/src/lib.rs

```diff
@@ -856,6 +868,30 @@ where
 				}
 
 				actions
+			},
+			(block_hash, validator_index, candidate_hash) = sign_approvals_timers.timers.select_next_some() => {
+				gum::debug!(
+					target: LOG_TARGET,
+					?candidate_hash,
+					"Sign approval for candidate",
+				);
+				if let Err(err) = maybe_create_signature(
+					&mut overlayed_db,
+					&mut session_info_provider,
+					&state, &mut ctx,
+					block_hash,
+					validator_index,
+					candidate_hash,
```

Comment:
Why do we still need candidate hash ? It could just pickup whatever is pending. We should also cleanup timers for approvals already signed and sent when max coalescing limit is reached for example.

Reply:
> Why do we still need candidate hash
Yeah that's for the cleanup part.  The invariants I'm trying to enforce.
1. We don't queue more than MAX_APPROVAL_COALESCE_COUNT.
2. No approve candidate waits for more than MAX_APPROVAL_COALESCE_WAIT.

The way that is achieved, is that we always start a MAX_APPROVAL_COALESCE_WAIT timer for all candidates and we create a signature only if the candidate for which we started the timer is still in the list of queued candidates, if it's not, some other timer already created a signature so it does nothing and let the next timer decide if it needs to create a signature based on the same rule.

That seems non-obvious, so I'll update it to use a single timer.